---
title: "Projet 8 - Master Regulators for Metastatic behavior in Osteosarcoma"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Chabrier Lisa, Cortes Garcia Estaban, Dorval LoÃ¯c

# I- Introduction
TODO : Lecture d'article, intro du sujet.

# II- Analysis
Project step by step:
- You should download the gene expression data generated by the study (GSE21257);

- You should install the RTN package and infer the regulatory network based on the down-
loaded gene expression data;

- You should install snow package for parallel processing in RTN;
- You should install RedeR package to better visualize the network you inferred;

- By performing a differential gene expression analysis between the metastatic and the non-
metastatic biopsies, you will obtain a signature for the metastatic behavior of this cancer

in these patients;
- You should run the Master Regulator Analysis to infer putative Master Regulators by using
the inferred network and the obtained signature.
- Get biological insight into these putative master regulators at the Human Protein Atlas
and at Gene Cards.

```{r}
library(rstudioapi) # Set working directory
library(RTN) # Infer regulatory network based on the downloaded gene expression data
library(snow) # Parallel processing in RTN
library(RedeR) # Better visualization of the network
library(CoRegNet) # Get the human TFs
library(lumi) # Convert nuID > entrezID
library(lumiHumanIDMapping) # Convert nuID > entrezID

data("HumanTF_entrezgene")

# Design matrix : G0 = no mestatases.

#### Getting the path of your current open file ####
current_path = rstudioapi::getActiveDocumentContext()$path 
setwd(dirname(current_path ))
print(getwd())
```

```{r}
################################################################
####   Loading & preprocessing the data  ####
################################################################
library(Biobase)
library(GEOquery)
library(limma)

#### load series and platform data from GEO ####

gset <- getGEO("GSE21257", GSEMatrix =TRUE, AnnotGPL=FALSE)
if (length(gset) > 1) idx <- grep("GPL10295", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

# make proper column names to match toptable 
fvarLabels(gset) <- make.names(fvarLabels(gset))

# group names for all samples
gsms <- "11111111111111111111100001111011010001010011110000000"
sml <- c()
for (i in 1:nchar(gsms)) { sml[i] <- substr(gsms,i,i) }

#### log2 transform ####
ex <- exprs(gset)
qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
LogC <- (qx[5] > 100) ||
  (qx[6]-qx[1] > 50 && qx[2] > 0) ||
  (qx[2] > 0 && qx[2] < 1 && qx[4] > 1 && qx[4] < 2)
if (LogC) { ex[which(ex <= 0)] <- NaN
exprs(gset) <- log2(ex) }
```

```{r}
################################################################
####    Network inference     ####
################################################################

#### Preprocessing ####
rowAnnotation <- as.matrix(nuID2EntrezID(nuID = rownames(ex), lib.mapping = 'lumiHumanIDMapping'))
rowAnnotation <- cbind(nuIDtoEntrez, rownames(nuIDtoEntrez))
colnames(rowAnnotation) <- c("entrez", "nuID")
rowAnnotation <- rowAnnotation[,c("nuID", "entrez")]

#### TFs from immune cell types of interest ####
macrophage <- c("ATM", "JUN", "CEBPB", "CREB1", "IRF1", "IRF4", "IRF5", 
                "IRF7", "IRF8", "MKI67", "MNDA", "NFKB1", "TP53", "PCNA", 
                "PPARG", "SPI1", "SSRP1", "STAT1", "STAT3", "STAT6", "XBP1")
macrophage <- mapIds(org.Hs.eg.db, macrophage, "ENTREZID", "SYMBOL")

Bcell <- c("BACH2", "BCL6", "PRDM1", "IRF4", "IRF8", "PAX5", "SPI1", 
           "STAT5", "XBP1")
Bcell <- mapIds(org.Hs.eg.db, Bcell, "ENTREZID", "SYMBOL")

Tcell <- c("IKZF3", "ATM", "BATF", "BCL6", "PRDM1", "FOXP3", "GATA3",
           "IKZF2", "IKZF1", "IRF4", "IRF7", "MKI67", "NFKB1", "TP53",
           "PCNA", "SSRP1", "STAT1", "STAT4", "TBX21", "TCL1", "ZBTB7B",
           "ZAP70")
Tcell <- mapIds(org.Hs.eg.db, Tcell, "ENTREZID", "SYMBOL")

#### Transcriptional Network Inference for the macrophage TFs ####
rtni_macro <- tni.constructor(expData = as.matrix(ex),
                        regulatoryElements = macrophage,
                        rowAnnotation = rowAnnotation)
rtni_macro <- tni.permutation(rtni_macro, nPermutations = 100)
rtni_macro <- tni.bootstrap(rtni_macro)
rtni_macro <- tni.dpi.filter(rtni_macro)
tni.regulon.summary(rtni_macro)
g_macro <- tni.graph(rtni_macro, tfs = macrophage)

# Visualization on RedeR
rdp <- RedPort()
calld(rdp, checkcalls=T)
addGraph(rdp, g_macro, layout=NULL)
addLegend.color(rdp, g_macro, type="edge")
addLegend.shape(rdp, g_macro)
relax(rdp, ps = TRUE)
```

```{r}
#### Transcriptional Network Inference for the B cell TFs####
rtni_Bcell <- tni.constructor(expData = as.matrix(ex),
                              regulatoryElements = Bcell,
                              rowAnnotation = rowAnnotation)
rtni_Bcell <- tni.permutation(rtni_Bcell, nPermutations = 100)
rtni_Bcell <- tni.bootstrap(rtni_Bcell)
rtni_Bcell <- tni.dpi.filter(rtni_Bcell)
tni.regulon.summary(rtni_Bcell)
g_Bcell <- tni.graph(rtni_Bcell, tfs = Bcell)

# Visualization on RedeR
rdp <- RedPort()
calld(rdp, checkcalls=T)
addGraph(rdp, g_Bcell, layout=NULL)
addLegend.color(rdp, g_Bcell, type="edge")
addLegend.shape(rdp, g_Bcell)
relax(rdp, ps = TRUE)
```

```{r}
#### Transcriptional Network Inference for the T cell TFs####
rtni_Tcell <- tni.constructor(expData = as.matrix(ex),
                              regulatoryElements = Tcell,
                              rowAnnotation = rowAnnotation)
rtni_Tcell <- tni.permutation(rtni_Tcell, nPermutations = 100)
rtni_Tcell <- tni.bootstrap(rtni_Tcell)
rtni_Tcell <- tni.dpi.filter(rtni_Tcell)
tni.regulon.summary(rtni_Tcell)
g_Tcell <- tni.graph(rtni_Tcell, tfs = Tcell)

# Visualization on RedeR
rdp <- RedPort()
calld(rdp, checkcalls=T)
addGraph(rdp, g_Tcell, layout=NULL)
addLegend.color(rdp, g_Tcell, type="edge")
addLegend.shape(rdp, g_Tcell)
relax(rdp, ps = TRUE)
```

```{r}
################################################################
####    Differential gene expression analysis   ####
################################################################

# set up the data and proceed with analysis
sml <- paste("G", sml, sep="")    # set group names
fl <- as.factor(sml)
gset$description <- fl
design <- model.matrix(~ description + 0, gset)
colnames(design) <- levels(fl)
fit <- lmFit(gset, design)
cont.matrix <- makeContrasts(G1-G0, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2, 0.01)
tT <- topTable(fit2, adjust="fdr", sort.by="B", number=250)

tT <- subset(tT, select=c("ID","adj.P.Val","P.Value","t","B","logFC","GB_ACC","SEQUENCE"))
write.table(tT, file=stdout(), row.names=F, sep="\t")

#### Volcano plots ####
volcanoplot(fit2, highlight = 15) # Improve quality of the figure

# Possibly add heatmaps, pathways analysis, GSEA ...
```

```{r}
################################################################
####    Transcriptional Network Analysis    ####
################################################################

tTb <- subset(tT, tT$B >= 1.9)
tTb.entrez <- nuID2EntrezID(tTb$ID, lib.mapping = 'lumiHumanIDMapping')
tTb <- cbind(tTb, tTb.entrez)
tT.entrez <- nuID2EntrezID(tT$ID, lib.mapping = 'lumiHumanIDMapping')
tT <- cbind(tT, tT.entrez)

b.stat <- tT$B
names(b.stat) <- rownames(tT)
logFC <- tT$logFC
names(logFC) <- rownames(tT)
tTb.hits <- rownames(tTb)
tTb.hits <- as.character(tTb.hits)
phenoID <- rownames(tT)
phenoID <- cbind(phenoID, tT.entrez)

###
### Modify rtna object using the rtni of interest
###
# Input 1: 'object', a TNI object with regulons
# Input 2: 'phenotype', a named numeric vector, usually log2 differential expression levels
# Input 3: 'hits', a character vector, usually a set of differentially expressed genes
# Input 4: 'phenoIDs', an optional data frame with gene anottation mapped to the phenotype
rtna <- tni2tna.preprocess(object = rtni,
                           phenotype = logFC,
                           hits = tTb.hits,
                           phenoIDs = phenoID) # problem here

# Run the MRA method
rtna <- tna.mra(rtna)

# Get MRA results;
#..setting 'ntop = -1' will return all results, regardless of a threshold
mra <- tna.get(rtna, what="mra", ntop = -1)
head(mra)

# Run the GSEA method
# Please set nPermutations >= 1000
rtna <- tna.gsea1(rtna, stepFilter=FALSE, nPermutations=100)

# Get GSEA results
gsea1 <- tna.get(rtna, what="gsea1", ntop = -1)
head(gsea1)

# Plot GSEA results
tna.plot.gsea1(rtna, labPheno="abs(log2 fold changes)", ntop = -1)

# Run the GSEA-2T method
# Please set nPermutations >= 1000
rtna <- tna.gsea2(rtna, stepFilter = FALSE, nPermutations = 100)

# Get GSEA-2T results
gsea2 <- tna.get(rtna, what = "gsea2", ntop = -1)
head(gsea2$differential)

# Plot GSEA-2T results
tna.plot.gsea2(rtna, labPheno="log2 fold changes", tfs="PTTG1")
```

```{r}
# Useful ?
################################################################
####   Boxplot for selected GEO samples   ####
################################################################
library(Biobase)
library(GEOquery)

# load series and platform data from GEO

gset <- getGEO("GSE21257", GSEMatrix =TRUE, getGPL=FALSE)
if (length(gset) > 1) idx <- grep("GPL10295", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

# group names for all samples in a series
gsms <- "11111111111111111111100001111011010001010011110000000"
sml <- c()
for (i in 1:nchar(gsms)) { sml[i] <- substr(gsms,i,i) }
sml <- paste("G", sml, sep="")  # set group names

# order samples by group
ex <- exprs(gset)[ , order(sml)]
sml <- sml[order(sml)]
fl <- as.factor(sml)
labels <- c("no+metastases","metastases")

# set parameters and draw the plot
palette(c("#f4dfdf","#dfeaf4", "#AABBCC"))
dev.new(width=4+dim(gset)[[2]]/5, height=6)
par(mar=c(2+round(max(nchar(sampleNames(gset)))/2),4,2,1))
title <- paste ("GSE21257", '/', annotation(gset), " selected samples", sep ='')
boxplot(ex, boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=fl)
legend("topleft", labels, fill=palette(), bty="n")
```

